clients/src/main/java/org/apache/kafka/common/internals/Plugin.java

```diff
@@ -40,14 +47,46 @@ private Plugin(T instance, PluginMetricsImpl pluginMetrics) {
         this.pluginMetrics = Optional.ofNullable(pluginMetrics);
     }
 
+    /**
+     * Wrap an instance into a Plugin.
+     * @param instance the instance to wrap
+     * @param metrics the metrics
+     * @param tagsSupplier supplier to retrieve the tags
+     * @return the plugin
+     */
+    public static <T> Plugin<T> wrapInstance(T instance, Metrics metrics, Supplier<Map<String, String>> tagsSupplier) {
+        PluginMetricsImpl pluginMetrics = null;
+        if (instance instanceof Monitorable && metrics != null) {
+            pluginMetrics = new PluginMetricsImpl(metrics, tagsSupplier.get());
+            ((Monitorable) instance).withPluginMetrics(pluginMetrics);
+        }
+        return new Plugin<>(instance, pluginMetrics);
+    }
```

Comment:
Also ask the method PluginMetricsImpl#metricName, it also use `Map.putAll()`, Should we also update it for order?

```
@Override
public MetricName metricName(String name, String description, Map<String, String> tags) {
    if (closing) throw new IllegalStateException("This PluginMetrics instance is closed");
    for (String tagName : tags.keySet()) {
        if (this.tags.containsKey(tagName)) {
            throw new IllegalArgumentException("Cannot use " + tagName + " as a tag name");
        }
    }
    Map<String, String> metricsTags = new LinkedHashMap<>(this.tags);
    metricsTags.putAll(tags);
    return metrics.metricName(name, GROUP, description, metricsTags);
}
```

Reply:
Maybe we should update the `PluginMetrics` API to require a `LinkedHashMap`, or at least mention it in the javadoc.

Reply:
I agree. I’d prefer to require a `LinkedHashMap` rather than just mentioning it in the JavaDoc, since JavaDoc alone doesn’t enforce anything.

Reply:
Feel free to open a jira if you want to handle this.

Reply:
Sure, Open https://issues.apache.org/jira/browse/KAFKA-19139 to trace it

Reply:
Thanks!